File: schema/runtime_test.go
package schema_test

import (
	"encoding/json"
	"os"
	"path/filepath"
	"testing"

	"github.com/grafana/dsconfig/schema"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

// ============================================================
// NewDatasourceConfig
// ============================================================

func TestNewDatasourceConfig_ParsesAllSections(t *testing.T) {
	raw := map[string]any{
		"url":       "https://example.com",
		"basicAuth": true,
		"jsonData": map[string]any{
			"timeout": float64(30),
		},
		"secureJsonData": map[string]any{
			"password": "secret",
		},
		"secureJsonFields": map[string]any{
			"password": true,
		},
	}

	dc := schema.NewDatasourceConfig(raw)

	assert.Equal(t, "https://example.com", dc.Root["url"])
	assert.Equal(t, true, dc.Root["basicAuth"])
	assert.Equal(t, float64(30), dc.JSONData["timeout"])
	assert.Equal(t, "secret", dc.SecureJSONData["password"])
	assert.True(t, dc.SecureJSONFields["password"])
}

// ============================================================
// LoadAndValidate — simple field extraction
// ============================================================

func TestLoadAndValidate_RootField(t *testing.T) {
	s := minimalSchema(schema.ConfigField{
		ID: "url", Key: "url", ValueType: schema.StringType,
		Target: ptr(schema.RootTarget), Required: true,
	})

	config := schema.DatasourceConfig{
		Root: map[string]any{"url": "https://example.com"},
	}

	result, err := schema.LoadAndValidate(s, config, schema.ReadMode)
	require.NoError(t, err)
	assert.False(t, result.HasErrors())
	assert.Equal(t, "https://example.com", result.Values["url"].Value)
	assert.Equal(t, schema.SourceConfig, result.Values["url"].Source)
}

func TestLoadAndValidate_JSONDataField(t *testing.T) {
	s := minimalSchema(schema.ConfigField{
		ID: "timeout", Key: "timeout", ValueType: schema.NumberType,
		Target: ptr(schema.JSONDataTarget),
	})

	config := schema.DatasourceConfig{
		JSONData: map[string]any{"timeout": float64(30)},
	}

	result, err := schema.LoadAndValidate(s, config, schema.ReadMode)
	require.NoError(t, err)
	assert.False(t, result.HasErrors())
	assert.Equal(t, float64(30), result.Values["timeout"].Value)
}

func TestLoadAndValidate_SectionField(t *testing.T) {
	s := minimalSchema(schema.ConfigField{
		ID: "oauth.clientId", Key: "clientId", ValueType: schema.StringType,
		Target: ptr(schema.JSONDataTarget), Section: "oauth2",
	})

	config := schema.DatasourceConfig{
		JSONData: map[string]any{
			"oauth2": map[string]any{
				"clientId": "my-client",
			},
		},
	}

	result, err := schema.LoadAndValidate(s, config, schema.ReadMode)
	require.NoError(t, err)
	assert.Equal(t, "my-client", result.Values["oauth.clientId"].Value)
}

func TestLoadAndValidate_DottedSectionField(t *testing.T) {
	s := minimalSchema(schema.ConfigField{
		ID: "oauth.tokenUrl", Key: "tokenUrl", ValueType: schema.StringType,
		Target: ptr(schema.JSONDataTarget), Section: "oauth2.endpoints",
	})

	config := schema.DatasourceConfig{
		JSONData: map[string]any{
			"oauth2": map[string]any{
				"endpoints": map[string]any{
					"tokenUrl": "https://auth.example.com/token",
				},
			},
		},
	}

	result, err := schema.LoadAndValidate(s, config, schema.ReadMode)
	require.NoError(t, err)
	assert.Equal(t, "https://auth.example.com/token", result.Values["oauth.tokenUrl"].Value)
}

// ============================================================
// Defaults
// ============================================================

func TestLoadAndValidate_AppliesDefault(t *testing.T) {
	s := minimalSchema(schema.ConfigField{
		ID: "timeout", Key: "timeout", ValueType: schema.NumberType,
		Target: ptr(schema.JSONDataTarget), DefaultValue: float64(30),
	})

	config := schema.DatasourceConfig{
		JSONData: map[string]any{},
	}

	result, err := schema.LoadAndValidate(s, config, schema.ReadMode)
	require.NoError(t, err)
	assert.Equal(t, float64(30), result.Values["timeout"].Value)
	assert.Equal(t, schema.SourceDefault, result.Values["timeout"].Source)
}

func TestLoadAndValidate_ConfigOverridesDefault(t *testing.T) {
	s := minimalSchema(schema.ConfigField{
		ID: "timeout", Key: "timeout", ValueType: schema.NumberType,
		Target: ptr(schema.JSONDataTarget), DefaultValue: float64(30),
	})

	config := schema.DatasourceConfig{
		JSONData: map[string]any{"timeout": float64(60)},
	}

	result, err := schema.LoadAndValidate(s, config, schema.ReadMode)
	require.NoError(t, err)
	assert.Equal(t, float64(60), result.Values["timeout"].Value)
	assert.Equal(t, schema.SourceConfig, result.Values["timeout"].Source)
}

// ============================================================
// Required validation
// ============================================================

func TestLoadAndValidate_RequiredFieldMissing(t *testing.T) {
	s := minimalSchema(schema.ConfigField{
		ID: "url", Key: "url", ValueType: schema.StringType,
		Target: ptr(schema.RootTarget), Required: true,
	})

	config := schema.DatasourceConfig{Root: map[string]any{}}

	result, err := schema.LoadAndValidate(s, config, schema.ReadMode)
	require.NoError(t, err)
	assert.True(t, result.HasErrors())
	assert.Equal(t, "required", result.Errors[0].Code)
	assert.Equal(t, "url", result.Errors[0].FieldID)
}

// ============================================================
// Type validation
// ============================================================

func TestLoadAndValidate_TypeMismatch(t *testing.T) {
	s := minimalSchema(schema.ConfigField{
		ID: "timeout", Key: "timeout", ValueType: schema.NumberType,
		Target: ptr(schema.JSONDataTarget),
	})

	config := schema.DatasourceConfig{
		JSONData: map[string]any{"timeout": "not-a-number"},
	}

	result, err := schema.LoadAndValidate(s, config, schema.ReadMode)
	require.NoError(t, err)
	assert.True(t, result.HasErrors())
	assert.Equal(t, "type_mismatch", result.Errors[0].Code)
}

// ============================================================
// Validation rules
// ============================================================

func TestLoadAndValidate_PatternValidation(t *testing.T) {
	s := minimalSchema(schema.ConfigField{
		ID: "url", Key: "url", ValueType: schema.StringType,
		Target: ptr(schema.RootTarget),
		Validations: []schema.FieldValidationRule{
			{Type: schema.PatternValidation, Pattern: "^https?://"},
		},
	})

	t.Run("valid", func(t *testing.T) {
		config := schema.DatasourceConfig{Root: map[string]any{"url": "https://example.com"}}
		result, err := schema.LoadAndValidate(s, config, schema.ReadMode)
		require.NoError(t, err)
		assert.False(t, result.HasErrors())
	})

	t.Run("invalid", func(t *testing.T) {
		config := schema.DatasourceConfig{Root: map[string]any{"url": "ftp://example.com"}}
		result, err := schema.LoadAndValidate(s, config, schema.ReadMode)
		require.NoError(t, err)
		assert.True(t, result.HasErrors())
		assert.Equal(t, "pattern", result.Errors[0].Code)
	})
}

func TestLoadAndValidate_RangeValidation(t *testing.T) {
	s := minimalSchema(schema.ConfigField{
		ID: "timeout", Key: "timeout", ValueType: schema.NumberType,
		Target: ptr(schema.JSONDataTarget),
		Validations: []schema.FieldValidationRule{
			{Type: schema.RangeValidation, Min: ptr(1.0), Max: ptr(300.0)},
		},
	})

	t.Run("in range", func(t *testing.T) {
		config := schema.DatasourceConfig{JSONData: map[string]any{"timeout": float64(30)}}
		result, _ := schema.LoadAndValidate(s, config, schema.ReadMode)
		assert.False(t, result.HasErrors())
	})

	t.Run("below min", func(t *testing.T) {
		config := schema.DatasourceConfig{JSONData: map[string]any{"timeout": float64(0)}}
		result, _ := schema.LoadAndValidate(s, config, schema.ReadMode)
		assert.True(t, result.HasErrors())
		assert.Equal(t, "range", result.Errors[0].Code)
	})

	t.Run("above max", func(t *testing.T) {
		config := schema.DatasourceConfig{JSONData: map[string]any{"timeout": float64(999)}}
		result, _ := schema.LoadAndValidate(s, config, schema.ReadMode)
		assert.True(t, result.HasErrors())
	})
}

func TestLoadAndValidate_AllowedValuesValidation(t *testing.T) {
	s := minimalSchema(schema.ConfigField{
		ID: "method", Key: "httpMethod", ValueType: schema.StringType,
		Target: ptr(schema.JSONDataTarget),
		Validations: []schema.FieldValidationRule{
			{Type: schema.AllowedValuesValidation, Values: []any{"GET", "POST"}},
		},
	})

	t.Run("allowed", func(t *testing.T) {
		config := schema.DatasourceConfig{JSONData: map[string]any{"httpMethod": "GET"}}
		result, _ := schema.LoadAndValidate(s, config, schema.ReadMode)
		assert.False(t, result.HasErrors())
	})

	t.Run("not allowed", func(t *testing.T) {
		config := schema.DatasourceConfig{JSONData: map[string]any{"httpMethod": "DELETE"}}
		result, _ := schema.LoadAndValidate(s, config, schema.ReadMode)
		assert.True(t, result.HasErrors())
		assert.Equal(t, "allowedValues", result.Errors[0].Code)
	})
}

func TestLoadAndValidate_ItemCountValidation(t *testing.T) {
	s := minimalSchema(schema.ConfigField{
		ID: "tags", Key: "tags", ValueType: schema.ArrayType,
		Target: ptr(schema.JSONDataTarget),
		Item:   &schema.FieldItemSchema{ValueType: schema.StringType},
		Validations: []schema.FieldValidationRule{
			{Type: schema.ItemCountValidation, Max: ptr(3.0)},
		},
	})

	t.Run("within limit", func(t *testing.T) {
		config := schema.DatasourceConfig{JSONData: map[string]any{"tags": []any{"a", "b"}}}
		result, _ := schema.LoadAndValidate(s, config, schema.ReadMode)
		assert.False(t, result.HasErrors())
	})

	t.Run("exceeds limit", func(t *testing.T) {
		config := schema.DatasourceConfig{JSONData: map[string]any{"tags": []any{"a", "b", "c", "d"}}}
		result, _ := schema.LoadAndValidate(s, config, schema.ReadMode)
		assert.True(t, result.HasErrors())
		assert.Equal(t, "itemCount", result.Errors[0].Code)
	})
}

// ============================================================
// Secure fields
// ============================================================

func TestLoadAndValidate_SecureField_ReadMode(t *testing.T) {
	s := minimalSchema(schema.ConfigField{
		ID: "password", Key: "password", ValueType: schema.StringType,
		Target: ptr(schema.SecureJSONTarget), Required: true,
	})

	t.Run("configured", func(t *testing.T) {
		config := schema.DatasourceConfig{
			SecureJSONFields: map[string]bool{"password": true},
		}
		result, _ := schema.LoadAndValidate(s, config, schema.ReadMode)
		assert.False(t, result.HasErrors())
		assert.Equal(t, schema.SecureConfigured, result.SecureFields["password"])
	})

	t.Run("missing required", func(t *testing.T) {
		config := schema.DatasourceConfig{
			SecureJSONFields: map[string]bool{},
		}
		result, _ := schema.LoadAndValidate(s, config, schema.ReadMode)
		assert.True(t, result.HasErrors())
		assert.Equal(t, schema.SecureUnset, result.SecureFields["password"])
	})
}

func TestLoadAndValidate_SecureField_WriteMode(t *testing.T) {
	s := minimalSchema(schema.ConfigField{
		ID: "password", Key: "password", ValueType: schema.StringType,
		Target: ptr(schema.SecureJSONTarget), Required: true,
	})

	t.Run("provided", func(t *testing.T) {
		config := schema.DatasourceConfig{
			SecureJSONData: map[string]any{"password": "s3cret"},
		}
		result, _ := schema.LoadAndValidate(s, config, schema.WriteMode)
		assert.False(t, result.HasErrors())
		assert.Equal(t, schema.SecureUpdated, result.SecureFields["password"])
		assert.Equal(t, "s3cret", result.Values["password"].Value)
	})

	t.Run("missing required", func(t *testing.T) {
		config := schema.DatasourceConfig{
			SecureJSONData: map[string]any{},
		}
		result, _ := schema.LoadAndValidate(s, config, schema.WriteMode)
		assert.True(t, result.HasErrors())
		assert.Equal(t, schema.SecureUnset, result.SecureFields["password"])
	})
}

// ============================================================
// Virtual fields are skipped
// ============================================================

func TestLoadAndValidate_SkipsVirtualFields(t *testing.T) {
	s := minimalSchema(
		schema.ConfigField{
			ID: "url", Key: "url", ValueType: schema.StringType,
			Target: ptr(schema.RootTarget),
		},
		schema.ConfigField{
			ID: "derived", Key: "derived", ValueType: schema.BooleanType,
			Kind: schema.VirtualField,
		},
	)

	config := schema.DatasourceConfig{Root: map[string]any{"url": "https://example.com"}}
	result, err := schema.LoadAndValidate(s, config, schema.ReadMode)
	require.NoError(t, err)
	assert.Contains(t, result.Values, "url")
	assert.NotContains(t, result.Values, "derived")
}

// ============================================================
// IndexedPair expansion
// ============================================================

func TestLoadAndValidate_IndexedPairExpansion(t *testing.T) {
	trueVal := true
	s := minimalSchema(schema.ConfigField{
		ID: "httpHeaders", Key: "httpHeaders", ValueType: schema.ArrayType,
		Target: ptr(schema.JSONDataTarget),
		Item: &schema.FieldItemSchema{
			ValueType: schema.ObjectType,
			Fields: []schema.ConfigField{
				{ID: "httpHeaders.item.name", Key: "name", ValueType: schema.StringType, IsItemField: &trueVal},
				{ID: "httpHeaders.item.value", Key: "value", ValueType: schema.StringType, IsItemField: &trueVal},
			},
		},
		Storage: &schema.StorageMapping{
			Type:       schema.IndexedPairMapping,
			Key:        &schema.MappingField{Target: schema.JSONDataTarget, Pattern: "httpHeaderName{index}"},
			Value:      &schema.MappingField{Target: schema.SecureJSONTarget, Pattern: "httpHeaderValue{index}"},
			StartIndex: ptr(1),
		},
	})

	config := schema.DatasourceConfig{
		JSONData: map[string]any{
			"httpHeaderName1": "X-Custom",
			"httpHeaderName2": "X-Token",
		},
		SecureJSONData: map[string]any{
			"httpHeaderValue1": "value1",
			"httpHeaderValue2": "value2",
		},
	}

	result, err := schema.LoadAndValidate(s, config, schema.WriteMode)
	require.NoError(t, err)
	assert.False(t, result.HasErrors())

	arr, ok := result.Values["httpHeaders"].Value.([]any)
	require.True(t, ok)
	require.Len(t, arr, 2)

	item0 := arr[0].(map[string]any)
	assert.Equal(t, "X-Custom", item0["name"])
	assert.Equal(t, "value1", item0["value"])

	item1 := arr[1].(map[string]any)
	assert.Equal(t, "X-Token", item1["name"])
	assert.Equal(t, "value2", item1["value"])
}

func TestLoadAndValidate_IndexedPairEmpty(t *testing.T) {
	trueVal := true
	s := minimalSchema(schema.ConfigField{
		ID: "httpHeaders", Key: "httpHeaders", ValueType: schema.ArrayType,
		Target: ptr(schema.JSONDataTarget),
		Item: &schema.FieldItemSchema{
			ValueType: schema.ObjectType,
			Fields: []schema.ConfigField{
				{ID: "httpHeaders.item.name", Key: "name", ValueType: schema.StringType, IsItemField: &trueVal},
				{ID: "httpHeaders.item.value", Key: "value", ValueType: schema.StringType, IsItemField: &trueVal},
			},
		},
		Storage: &schema.StorageMapping{
			Type:       schema.IndexedPairMapping,
			Key:        &schema.MappingField{Target: schema.JSONDataTarget, Pattern: "httpHeaderName{index}"},
			Value:      &schema.MappingField{Target: schema.SecureJSONTarget, Pattern: "httpHeaderValue{index}"},
			StartIndex: ptr(1),
		},
	})

	config := schema.DatasourceConfig{
		JSONData: map[string]any{},
	}

	result, err := schema.LoadAndValidate(s, config, schema.ReadMode)
	require.NoError(t, err)
	assert.Nil(t, result.Values["httpHeaders"].Value)
	assert.Equal(t, schema.SourceNone, result.Values["httpHeaders"].Source)
}

// ============================================================
// File-based integration tests using existing testdata
// ============================================================

func TestLoadAndValidate_FromTestdata(t *testing.T) {
	cases := []struct {
		dir      string
		mode     schema.LoadMode
		hasError bool
	}{
		{dir: "simple-url", mode: schema.ReadMode},
		{dir: "root-jsondata-secure-mix", mode: schema.WriteMode},
		{dir: "bearer-token-auth", mode: schema.WriteMode},
		{dir: "indexed-headers-storage", mode: schema.WriteMode},
		{dir: "nested-object-jsondata", mode: schema.ReadMode},
	}

	for _, tc := range cases {
		t.Run(tc.dir, func(t *testing.T) {
			dir := filepath.Join("testdata", "convert", tc.dir)

			inputBytes, err := os.ReadFile(filepath.Join(dir, "input.json"))
			require.NoError(t, err)

			configBytes, err := os.ReadFile(filepath.Join(dir, "config.json"))
			require.NoError(t, err)

			var s schema.DatasourceConfigSchema
			require.NoError(t, json.Unmarshal(inputBytes, &s))

			var rawConfig map[string]any
			require.NoError(t, json.Unmarshal(configBytes, &rawConfig))
			delete(rawConfig, "_comment")

			config := schema.NewDatasourceConfig(rawConfig)

			result, err := schema.LoadAndValidate(&s, config, tc.mode)
			require.NoError(t, err)

			if tc.hasError {
				assert.True(t, result.HasErrors(), "expected errors for %s", tc.dir)
			} else {
				assert.False(t, result.HasErrors(),
					"unexpected errors for %s: %v", tc.dir, result.Errors)
			}

			// Verify at least some values were extracted
			assert.NotEmpty(t, result.Values, "expected values for %s", tc.dir)
		})
	}
}

File: schema/runtime.test.ts
import { describe, it, expect } from "vitest"
import { readFileSync } from "fs"
import { join } from "path"
import type { DatasourceConfigSchema, ConfigField } from "./schema"
import {
    loadAndValidate,
    newDatasourceConfig,
    type DatasourceConfig,
    type LoadMode,
} from "./runtime"

// ============================================================
// Helpers
// ============================================================

function makeSchema(...fields: ConfigField[]): DatasourceConfigSchema {
    return {
        schemaVersion: "v1",
        pluginType: "test",
        pluginName: "Test",
        fields,
    }
}

// ============================================================
// newDatasourceConfig
// ============================================================

describe("newDatasourceConfig", () => {
    it("parses all sections", () => {
        const dc = newDatasourceConfig({
            url: "https://example.com",
            basicAuth: true,
            jsonData: { timeout: 30 },
            secureJsonData: { password: "secret" },
            secureJsonFields: { password: true },
        })

        expect(dc.root.url).toBe("https://example.com")
        expect(dc.root.basicAuth).toBe(true)
        expect(dc.jsonData?.timeout).toBe(30)
        expect(dc.secureJsonData?.password).toBe("secret")
        expect(dc.secureJsonFields?.password).toBe(true)
    })
})

// ============================================================
// Simple field extraction
// ============================================================

describe("loadAndValidate — extraction", () => {
    it("extracts root field", () => {
        const s = makeSchema({
            id: "url", key: "url", valueType: "string",
            target: "root", required: true,
        })
        const config: DatasourceConfig = {
            root: { url: "https://example.com" },
        }

        const result = loadAndValidate(s, config, "read")
        expect(result.errors).toHaveLength(0)
        expect(result.values.url.value).toBe("https://example.com")
        expect(result.values.url.source).toBe("config")
    })

    it("extracts jsonData field", () => {
        const s = makeSchema({
            id: "timeout", key: "timeout", valueType: "number",
            target: "jsonData",
        })
        const config: DatasourceConfig = {
            root: {},
            jsonData: { timeout: 30 },
        }

        const result = loadAndValidate(s, config, "read")
        expect(result.errors).toHaveLength(0)
        expect(result.values.timeout.value).toBe(30)
    })

    it("extracts section field", () => {
        const s = makeSchema({
            id: "oauth.clientId", key: "clientId", valueType: "string",
            target: "jsonData", section: "oauth2",
        })
        const config: DatasourceConfig = {
            root: {},
            jsonData: { oauth2: { clientId: "my-client" } },
        }

        const result = loadAndValidate(s, config, "read")
        expect(result.values["oauth.clientId"].value).toBe("my-client")
    })

    it("extracts dotted section field", () => {
        const s = makeSchema({
            id: "oauth.tokenUrl", key: "tokenUrl", valueType: "string",
            target: "jsonData", section: "oauth2.endpoints",
        })
        const config: DatasourceConfig = {
            root: {},
            jsonData: {
                oauth2: { endpoints: { tokenUrl: "https://auth.example.com/token" } },
            },
        }

        const result = loadAndValidate(s, config, "read")
        expect(result.values["oauth.tokenUrl"].value).toBe("https://auth.example.com/token")
    })
})

// ============================================================
// Defaults
// ============================================================

describe("loadAndValidate — defaults", () => {
    it("applies default when value missing", () => {
        const s = makeSchema({
            id: "timeout", key: "timeout", valueType: "number",
            target: "jsonData", defaultValue: 30,
        })
        const config: DatasourceConfig = { root: {}, jsonData: {} }

        const result = loadAndValidate(s, config, "read")
        expect(result.values.timeout.value).toBe(30)
        expect(result.values.timeout.source).toBe("default")
    })

    it("config overrides default", () => {
        const s = makeSchema({
            id: "timeout", key: "timeout", valueType: "number",
            target: "jsonData", defaultValue: 30,
        })
        const config: DatasourceConfig = { root: {}, jsonData: { timeout: 60 } }

        const result = loadAndValidate(s, config, "read")
        expect(result.values.timeout.value).toBe(60)
        expect(result.values.timeout.source).toBe("config")
    })
})

// ============================================================
// Required validation
// ============================================================

describe("loadAndValidate — required", () => {
    it("errors on missing required field", () => {
        const s = makeSchema({
            id: "url", key: "url", valueType: "string",
            target: "root", required: true,
        })
        const config: DatasourceConfig = { root: {} }

        const result = loadAndValidate(s, config, "read")
        expect(result.errors).toHaveLength(1)
        expect(result.errors[0].code).toBe("required")
        expect(result.errors[0].fieldId).toBe("url")
    })
})

// ============================================================
// Type validation
// ============================================================

describe("loadAndValidate — type check", () => {
    it("errors on type mismatch", () => {
        const s = makeSchema({
            id: "timeout", key: "timeout", valueType: "number",
            target: "jsonData",
        })
        const config: DatasourceConfig = {
            root: {},
            jsonData: { timeout: "not-a-number" },
        }

        const result = loadAndValidate(s, config, "read")
        expect(result.errors).toHaveLength(1)
        expect(result.errors[0].code).toBe("type_mismatch")
    })
})

// ============================================================
// Validation rules
// ============================================================

describe("loadAndValidate — validation rules", () => {
    it("pattern — valid", () => {
        const s = makeSchema({
            id: "url", key: "url", valueType: "string", target: "root",
            validations: [{ type: "pattern", pattern: "^https?://" }],
        })
        const result = loadAndValidate(s, { root: { url: "https://x.com" } }, "read")
        expect(result.errors).toHaveLength(0)
    })

    it("pattern — invalid", () => {
        const s = makeSchema({
            id: "url", key: "url", valueType: "string", target: "root",
            validations: [{ type: "pattern", pattern: "^https?://" }],
        })
        const result = loadAndValidate(s, { root: { url: "ftp://x.com" } }, "read")
        expect(result.errors).toHaveLength(1)
        expect(result.errors[0].code).toBe("pattern")
    })

    it("range — below min", () => {
        const s = makeSchema({
            id: "t", key: "t", valueType: "number", target: "jsonData",
            validations: [{ type: "range", min: 1, max: 300 }],
        })
        const result = loadAndValidate(s, { root: {}, jsonData: { t: 0 } }, "read")
        expect(result.errors).toHaveLength(1)
        expect(result.errors[0].code).toBe("range")
    })

    it("allowedValues — valid", () => {
        const s = makeSchema({
            id: "m", key: "m", valueType: "string", target: "jsonData",
            validations: [{ type: "allowedValues", values: ["GET", "POST"] }],
        })
        const result = loadAndValidate(s, { root: {}, jsonData: { m: "GET" } }, "read")
        expect(result.errors).toHaveLength(0)
    })

    it("allowedValues — invalid", () => {
        const s = makeSchema({
            id: "m", key: "m", valueType: "string", target: "jsonData",
            validations: [{ type: "allowedValues", values: ["GET", "POST"] }],
        })
        const result = loadAndValidate(s, { root: {}, jsonData: { m: "DELETE" } }, "read")
        expect(result.errors).toHaveLength(1)
        expect(result.errors[0].code).toBe("allowedValues")
    })

    it("itemCount — exceeds max", () => {
        const s = makeSchema({
            id: "tags", key: "tags", valueType: "array", target: "jsonData",
            item: { valueType: "string" },
            validations: [{ type: "itemCount", max: 3 }],
        })
        const result = loadAndValidate(s, { root: {}, jsonData: { tags: ["a", "b", "c", "d"] } }, "read")
        expect(result.errors).toHaveLength(1)
        expect(result.errors[0].code).toBe("itemCount")
    })
})

// ============================================================
// Secure fields
// ============================================================

describe("loadAndValidate — secure fields", () => {
    it("read mode — configured", () => {
        const s = makeSchema({
            id: "password", key: "password", valueType: "string",
            target: "secureJsonData", required: true,
        })
        const config: DatasourceConfig = {
            root: {},
            secureJsonFields: { password: true },
        }

        const result = loadAndValidate(s, config, "read")
        expect(result.errors).toHaveLength(0)
        expect(result.secureFields.password).toBe("configured")
    })

    it("read mode — missing required", () => {
        const s = makeSchema({
            id: "password", key: "password", valueType: "string",
            target: "secureJsonData", required: true,
        })
        const config: DatasourceConfig = { root: {} }

        const result = loadAndValidate(s, config, "read")
        expect(result.errors).toHaveLength(1)
        expect(result.secureFields.password).toBe("unset")
    })

    it("write mode — provided", () => {
        const s = makeSchema({
            id: "password", key: "password", valueType: "string",
            target: "secureJsonData", required: true,
        })
        const config: DatasourceConfig = {
            root: {},
            secureJsonData: { password: "s3cret" },
        }

        const result = loadAndValidate(s, config, "write")
        expect(result.errors).toHaveLength(0)
        expect(result.secureFields.password).toBe("updated")
        expect(result.values.password.value).toBe("s3cret")
    })

    it("write mode — missing required", () => {
        const s = makeSchema({
            id: "password", key: "password", valueType: "string",
            target: "secureJsonData", required: true,
        })
        const config: DatasourceConfig = { root: {}, secureJsonData: {} }

        const result = loadAndValidate(s, config, "write")
        expect(result.errors).toHaveLength(1)
        expect(result.secureFields.password).toBe("unset")
    })
})

// ============================================================
// Virtual fields
// ============================================================

describe("loadAndValidate — virtual fields", () => {
    it("skips virtual fields", () => {
        const s = makeSchema(
            { id: "url", key: "url", valueType: "string", target: "root" },
            { id: "derived", key: "derived", valueType: "boolean", kind: "virtual" },
        )
        const config: DatasourceConfig = { root: { url: "https://x.com" } }

        const result = loadAndValidate(s, config, "read")
        expect(result.values.url).toBeDefined()
        expect(result.values.derived).toBeUndefined()
    })
})

// ============================================================
// IndexedPair expansion
// ============================================================

describe("loadAndValidate — indexedPair", () => {
    function headersSchema(): DatasourceConfigSchema {
        return makeSchema({
            id: "httpHeaders", key: "httpHeaders", valueType: "array",
            target: "jsonData",
            item: {
                valueType: "object",
                fields: [
                    { id: "httpHeaders.item.name", key: "name", valueType: "string", isItemField: true },
                    { id: "httpHeaders.item.value", key: "value", valueType: "string", isItemField: true },
                ],
            },
            storage: {
                type: "indexedPair",
                key: { target: "jsonData", pattern: "httpHeaderName{index}" },
                value: { target: "secureJsonData", pattern: "httpHeaderValue{index}" },
                startIndex: 1,
            },
        })
    }

    it("expands indexed pairs", () => {
        const config: DatasourceConfig = {
            root: {},
            jsonData: { httpHeaderName1: "X-Custom", httpHeaderName2: "X-Token" },
            secureJsonData: { httpHeaderValue1: "val1", httpHeaderValue2: "val2" },
        }

        const result = loadAndValidate(headersSchema(), config, "write")
        expect(result.errors).toHaveLength(0)

        const arr = result.values.httpHeaders.value as unknown[]
        expect(arr).toHaveLength(2)
        expect((arr[0] as Record<string, unknown>).name).toBe("X-Custom")
        expect((arr[0] as Record<string, unknown>).value).toBe("val1")
        expect((arr[1] as Record<string, unknown>).name).toBe("X-Token")
        expect((arr[1] as Record<string, unknown>).value).toBe("val2")
    })

    it("handles empty indexed pairs", () => {
        const config: DatasourceConfig = { root: {}, jsonData: {} }

        const result = loadAndValidate(headersSchema(), config, "read")
        expect(result.values.httpHeaders.value).toBeNull()
        expect(result.values.httpHeaders.source).toBe("none")
    })
})

// ============================================================
// File-based integration tests
// ============================================================

describe("loadAndValidate — testdata integration", () => {
    const cases: Array<{ dir: string; mode: LoadMode }> = [
        { dir: "simple-url", mode: "read" },
        { dir: "root-jsondata-secure-mix", mode: "write" },
        { dir: "bearer-token-auth", mode: "write" },
        { dir: "indexed-headers-storage", mode: "write" },
        { dir: "nested-object-jsondata", mode: "read" },
    ]

    for (const tc of cases) {
        it(`${tc.dir}`, () => {
            const dir = join(__dirname, "testdata", "convert", tc.dir)
            const inputSchema: DatasourceConfigSchema = JSON.parse(
                readFileSync(join(dir, "input.json"), "utf-8"),
            )

            const rawConfig = JSON.parse(
                readFileSync(join(dir, "config.json"), "utf-8"),
            )
            delete rawConfig._comment

            const config = newDatasourceConfig(rawConfig)
            const result = loadAndValidate(inputSchema, config, tc.mode)

            expect(result.errors, `unexpected errors in ${tc.dir}: ${JSON.stringify(result.errors)}`).toHaveLength(0)
            expect(Object.keys(result.values).length).toBeGreaterThan(0)
        })
    }
})
